BoTorch fixes and test

sighellan opened the pull request:
*Fixes #522 *

*Description of changes:*
Changes:
- Set `torch.random.manual_seed()`: this makes the results reproducible for `n_workers=1`.
- Catch `ModelFittingError`: this fixes the scheduler sometimes crashing.
- Added test: check that the scheduler stays working.
- Update BoTorch requirement to 0.7.2
- Update setup.py so BoTorch requirements only gets installed for Python >= 3.8.
- Added links to BoTorch implementations in Readme where BoTorch is mentioned. State that Python >= 3.8 is required.
- Changes to tests so they only run on BoTorch is Python version is >= 3.8. 


*Edits*
- Test no longer checks for reproducibility. 
- Added changes after 'Added test'.

**New output run 1**
```python
--------------------
Resource summary (last result is reported):
 trial_id     status  iter  steps  width  height  step  mean_loss  epoch  worker-time
        0  Completed     5      5     10       0   4.0   2.000000    5.0     0.525761
        1  Completed     5      5     14      68   4.0   8.315152    5.0     0.530033
        2  Completed     5      5     18      23   4.0   3.519512    5.0     0.525702
        3  Completed     5      5     10      59   4.0   7.900000    5.0     0.532092
        4  Completed     5      5      8     100   4.0  12.380952    5.0     0.519296
        5  Completed     5      5     20     -73   4.0  -6.188889    5.0     0.524719
        6 InProgress     0      5     20    -100     -          -      -            -
1 trials running, 6 finished (6 until the end), 9.57s wallclock-time

mean_loss: best -6.18888888888889 for trial-id 5
--------------------
```

**New output run 2**
```python
--------------------
Resource summary (last result is reported):
 trial_id     status  iter  steps  width  height  step  mean_loss  epoch  worker-time
        0  Completed     5      5     10       0   4.0   2.000000    5.0     0.516101
        1  Completed     5      5     14      68   4.0   8.315152    5.0     0.523241
        2  Completed     5      5     18      23   4.0   3.519512    5.0     0.528620
        3  Completed     5      5     10      59   4.0   7.900000    5.0     0.522834
        4  Completed     5      5      8     100   4.0  12.380952    5.0     0.519298
        5  Completed     5      5     20     -73   4.0  -6.188889    5.0     0.523501
        6 InProgress     0      5     20    -100     -          -      -            -
1 trials running, 6 finished (6 until the end), 9.49s wallclock-time

mean_loss: best -6.18888888888889 for trial-id 5
--------------------
```

If we comment out the code to catch `ModelFittingError`we get
**Output ModelFittingError**
```python
--------------------
Resource summary (last result is reported):
 trial_id    status  iter  steps  width  height  step  mean_loss  epoch  worker-time
        0 Completed     5      5     10       0     4   2.000000      5     0.536383
        1 Completed     5      5     14      68     4   8.315152      5     0.522470
        2 Completed     5      5     18      23     4   3.519512      5     0.517027
0 trials running, 3 finished (3 until the end), 4.36s wallclock-time

mean_loss: best 2.0 for trial-id 0
--------------------

---------------------------------------------------------------------------
ModelFittingError                         Traceback (most recent call last)
Cell In[1], line 26
     15 # Set up tuner and run for a few evaluations
     16 tuner = Tuner(
     17         trial_backend=LocalBackend(entry_point=entry_point),
     18         scheduler=BoTorch(
   (...)
     24         sleep_time=0.001,
     25 )
---> 26 tuner.run()

File ~/[...]/syne_tune/tuner.py:296, in Tuner.run(self)
    291 except Exception as e:
    292     logger.error(
    293         "An error happened during the tuning, cleaning up resources and logging final resources "
    294         "before throwing the exception."
    295     )
--> 296     raise e
    297 finally:
    298     # graceful termination block called when the tuner reached its stop condition, when an error happened or
    299     # when the job got interrupted (can happen in spot-instances or when sending a SIGINT signal with ctrl+C).
    300     # the block displays the best configuration found and stops trials that may still be running.
    301     print_best_metric_found(
    302         tuning_status=self.tuning_status,
    303         metric_names=self.scheduler.metric_names(),
    304         mode=self.scheduler.metric_mode(),
    305     )

File ~/[...]/syne_tune/tuner.py:275, in Tuner.run(self)
    273 else:
    274     try:
--> 275         self._schedule_new_tasks(running_trials_ids=running_trials_ids)
    276     except StopIteration:
    277         logger.info(
    278             "Tuning is finishing as the whole configuration space got exhausted."
    279         )

File ~/[...]/syne_tune/tuner.py:458, in Tuner._schedule_new_tasks(self, running_trials_ids)
    456 # Schedule as many trials as we have free workers
    457 for _ in range(self.n_workers - num_busy_workers):
--> 458     trial = self._schedule_new_task()
    459     trial_id = trial.trial_id
    460     running_trials_ids.add(trial_id)

File ~/[...]/syne_tune/tuner.py:474, in Tuner._schedule_new_task(self)
    467 def _schedule_new_task(self) -> Optional[TrialResult]:
    468     """Schedules a new task according to scheduler suggestion.
    469 
    470     :return: Information for the trial suggested, ``None`` if the scheduler does
    471         not suggest a new configuration (this can happen if its configuration
    472         space is exhausted)
    473     """
--> 474     suggestion = self.scheduler.suggest(trial_id=self.trial_backend.new_trial_id())
    475     if suggestion is None:
    476         logger.info("Searcher ran out of candidates, tuning job is stopping.")

File ~/[...]/syne_tune/optimizer/schedulers/scheduler_searcher.py:62, in TrialSchedulerWithSearcher.suggest(self, trial_id)
     60 def suggest(self, trial_id: int) -> Optional[TrialSuggestion]:
     61     self._initialize_searcher()
---> 62     return super().suggest(trial_id)

File ~/[...]/syne_tune/optimizer/scheduler.py:142, in TrialScheduler.suggest(self, trial_id)
    113 def suggest(self, trial_id: int) -> Optional[TrialSuggestion]:
    114     """Returns a suggestion for a new trial, or one to be resumed
    115 
    116     This method returns ``suggestion`` of type :class:`TrialSuggestion` (unless
   (...)
    140         above. If no suggestion can be made, None is returned
    141     """
--> 142     ret_val = self._suggest(trial_id)
    143     if ret_val is not None:
    144         assert isinstance(ret_val, TrialSuggestion)

File ~/[...]/syne_tune/optimizer/schedulers/fifo.py:249, in FIFOScheduler._suggest(self, trial_id)
    247 extra_kwargs["elapsed_time"] = self._elapsed_time()
    248 trial_id = str(trial_id)
--> 249 config = self.searcher.get_config(**extra_kwargs, trial_id=trial_id)
    250 if config is not None:
    251     config = cast_config_values(config, self.config_space)

File ~/[...]/syne_tune/optimizer/schedulers/searchers/searcher.py:513, in SearcherWithRandomSeedAndFilterDuplicates.get_config(self, **kwargs)
    512 def get_config(self, **kwargs) -> Optional[dict]:
--> 513     new_config = self._get_config(**kwargs)
    514     if not self._allow_duplicates and new_config is not None:
    515         self._excl_list.add(new_config)

File ~/[...]/syne_tune/optimizer/schedulers/searchers/botorch/botorch_searcher.py:125, in BoTorchSearcher._get_config(self, trial_id, **kwargs)
    123         config_suggested = self._get_random_config()
    124     else:
--> 125         config_suggested = self._sample_next_candidate()
    127 if config_suggested is not None:
    128     self.trial_configs[trial_id] = config_suggested

File ~/[...]/syne_tune/optimizer/schedulers/searchers/botorch/botorch_searcher.py:194, in BoTorchSearcher._sample_next_candidate(self)
    192 gp = self._make_gp(X_tensor=X_tensor, Y_tensor=Y_tensor)
    193 mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
--> 194 fit_gpytorch_mll(mll, max_attempts=1)
    196 if self.pending_trials and self.fantasising and not subsample:
    197     X_pending = self._config_to_feature_matrix(self._configs_pending())

File ~/miniconda3/envs/[...]/lib/python3.8/site-packages/botorch/fit.py:105, in fit_gpytorch_mll(mll, closure, optimizer, closure_kwargs, optimizer_kwargs, **kwargs)
    102 if optimizer is not None:  # defer to per-method defaults
    103     kwargs["optimizer"] = optimizer
--> 105 return FitGPyTorchMLL(
    106     mll,
    107     type(mll.likelihood),
    108     type(mll.model),
    109     closure=closure,
    110     closure_kwargs=closure_kwargs,
    111     optimizer_kwargs=optimizer_kwargs,
    112     **kwargs,
    113 )

File ~/miniconda3/envs/[...]/lib/python3.8/site-packages/botorch/utils/dispatcher.py:93, in Dispatcher.__call__(self, *args, **kwargs)
     91 func = self.__getitem__(types=types)
     92 try:
---> 93     return func(*args, **kwargs)
     94 except MDNotImplementedError:
     95     # Traverses registered methods in order, yields whenever a match is found
     96     funcs = self.dispatch_iter(*types)

File ~/miniconda3/envs/[...]/lib/python3.8/site-packages/botorch/fit.py:283, in _fit_fallback(mll, _, __, closure, optimizer, closure_kwargs, optimizer_kwargs, max_attempts, warning_handler, caught_exception_types, **ignore)
    280 if debug.off():
    281     msg = msg + " For more information, try enabling botorch.settings.debug mode."
--> 283 raise ModelFittingError(msg)

ModelFittingError: All attempts to fit the model have failed. For more information, try enabling botorch.settings.debug mode.
```

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

mseeger commented on tst/schedulers/test_botorch.py:
```diff
@@ -0,0 +1,60 @@
+
+
+@pytest.mark.timeout(30)
+def test_botorch_reproducible():
```
I'd prefer not to have this test, unless BoTorch explicitly guarantees reproducible outputs. For our built-in
BO schedulers, we do not, and I am almost sure they do not either.

I'd like to avoid having tests that fail later for random reasons.

sighellan replied:
https://github.com/pytorch/botorch/issues/1190 They expect the results to be reproducible.

sighellan replied:
I understand the hesitation, though we seem to be able to make it reproducible now, so at least then this test would let us know when that changes, and if it's not on syne-tune's end it can just be removed then. 

I think there's a stronger argument for having some test that makes sure the scheduler runs, whether it needs to check for reproducibility is less important. For debugging things like `ModelFittingError`it helps to have the code be reproducible, since it only occurs rarely otherwise.

geoalgo replied:
Thanks for adapting the test! Could update the name which is now outdated? (for instance to `test_smoke_botorch`)

sighellan commented:
I am not sure why the test is failing for Python 3.7. It seems to be struggling with the botorch imports:

```python
110:  >       return normalize(X, bounds)
111: E       NameError: name 'normalize' is not defined
```

```python 
228: >       except NotPSDError as _:
229: E       NameError: name 'NotPSDError' is not defined
```

https://github.com/awslabs/syne-tune/actions/runs/4046593772/jobs/6959568576

sighellan commented:
Ok, so the issue is that while we're saying `botorch>=0.6.2` in [requirements-botorch.txt](https://github.com/awslabs/syne-tune/blob/main/requirements-botorch.txt), we're using `fit_gpytoch_mll`which wasn't introduced until `0.7.2`. But `0.7.2` no longer supports Python 3.7. 

The options seem to either be to change to `fit_gpytorch_model`instead of `fit_gpytorch_mll` (and also remove `ModelFittingError`which also wasn't around at the time), or to require Python 3.8 or later for the `BoTorchSearcher`.

mseeger commented:
I'd be in favour of requiring Python 3.8 for `BoTorchSearcher`.

Is there a way to only run a unit test if Python is >= 3.8?

sighellan commented:
> I'd be in favour of requiring Python 3.8 for `BoTorchSearcher`.
> 
> Is there a way to only run a unit test if Python is >= 3.8?

Yes we can skip it using `pytest.mark.skipif(sys.version_info <= (3, 7)`. But we also need to update the installation setup.

sighellan commented:
The tests are back to working after
-  updating setup.py so BoTorch only gets installed for Python >= 3.8
- making sure BoTorch gets skipped from other tests for Python 3.7

I've changed the new test to not check reproducibility, just whether the model runs, and updated the list of changes at the top of the pull request.

geoalgo commented on README.md:
```diff
@@ -136,7 +136,8 @@ RUSH | Zappella, et al. (2021)| random | yes | yes | yes
 The searchers fall into four broad categories, **deterministic**, **random**, **evolutionary** and **model-based**. The random searchers sample candidate hyperparameter configurations uniformly at random, while the model-based searchers sample them non-uniformly at random, according to a model (e.g., Gaussian process, density ration estimator, etc.) and an acquisition function. The evolutionary s ...
 
-Syne Tune also supports [BoTorch](https://github.com/pytorch/botorch) searchers.
+Syne Tune also supports [BoTorch](https://github.com/pytorch/botorch) searchers, with [an example](https://github.com/awslabs/syne-tune/blob/main/syne_tune/optimizer/schedulers/searchers/botorch/botorch_searcher.py) used by the [BoTorch scheduler](https://github.com/awslabs/syne-tune/blob/main/syne_tune/optimizer/baselines.py).
+Note that this requires Python >= 3.8.
```
```suggestion
Syne Tune also supports [BoTorch](https://github.com/awslabs/syne-tune/blob/main/syne_tune/optimizer/schedulers/searchers/botorch/botorch_searcher.py) searchers.
```

geoalgo commented on setup.py:
```diff
@@ -45,11 +46,14 @@ def load_benchmark_requirements():
     + required_yahpo
 )
+
+if sys.version_info >= (3, 8):
+    required_extra += required_botorch
```
```suggestion
# Botorch only supports python version > 3.8
if sys.version_info >= (3, 8):
    required_extra += required_botorch
```

geoalgo commented on tst/schedulers/test_botorch.py:
```diff
@@ -0,0 +1,56 @@
+    from syne_tune.optimizer.baselines import BoTorch
+
+    # Use train_height backend for our tests
+    entry_point = "examples/training_scripts/height_example/train_height.py"
```
This path would break if you run from a different directory, can you use something like this to avoid that side-effect:
https://github.com/awslabs/syne-tune/blob/main/examples/launch_height_baselines.py#L39

geoalgo reviewed (approved):
Thanks for this great fix!

mseeger commented:
Yes, thanks a lot. Hope this lets you progress with your project

sighellan commented:
Thank you!
